Add banner prop to EuiFlyoutBody

andreadelrio opened the pull request:
### Summary

In order to build the drilldowns feature, we'd like have a borderless callout at the top of a flyout. For that reason this PR:

- Adds a prop called `banner` to `EuiFlyOutBody` which takes a node
- Adds a prop called `borderless` to `EuiCallOut` which removes the usual border from callouts

![image](https://user-images.githubusercontent.com/4016496/74010728-fe346980-493a-11ea-83be-7a879055f965.png)

<img width="838" alt="Screen Shot 2020-02-06 at 10 17 34 PM" src="https://user-images.githubusercontent.com/4016496/74010778-1dcb9200-493b-11ea-9ccb-995d9615f351.png">

![flyout-callout](https://user-images.githubusercontent.com/4016496/74010978-8dda1800-493b-11ea-9a5d-4eaf4f6212e6.gif)



### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- [x] Checked in **mobile**
- [X] Checked in **IE11** and **Firefox**
- [X] Props have proper **autodocs**
- [x] Added **documentation** examples
- [x] Added or updated **jest tests**
~- [ ] Checked for **breaking changes** and labeled appropriately~
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- [x] A [changelog](https://github.com/elastic/eui/blob/master/CHANGELOG.md) entry exists and is marked appropriately

cchaos commented:
Hmm, part of this PR worries me a bit. Having two different styles of callouts (bordered and not) will cause some major inconsistencies in our applications. Mainly I'm thinking because devs won't necessarily know which one they should choose and pick entirely on aesthetics (which really is the only reason there's a difference). 

Looking at the flyout version, it does look nice and I can think of several places that would benefit from this design. Which brings up another thing to consider. Should those other places get updated to reflect this new design (again for consistency)? 

So my main question is, **does the need for the callout to extend to the sides of the flyout** (which is just an aesthetic choice at the moment) **outweigh the possibilities of inconsistent callout types throughout the applications**?

----

There's also another approach you could take that would ease the concern of inconsistent callouts, in which you apply a custom className to the provided `banner`, which in this case would be an `EuiCallout` and only when that banner className is used in conjunction with `.euiCallout` do you then remove the border. This would keep stand alone callout's always the same but only in this very specific instance would it remove the border.

----

If we do move forward with this, something to consider is that the overflow mask in the flyout body is causing the top portion of the callout to fade out as it approaches the header border. 

<img width="584" alt="Screen Shot 2020-02-07 at 08 46 55 AM" src="https://user-images.githubusercontent.com/549577/74034503-68bec700-4986-11ea-838f-62f70469aba6.png">

andreadelrio commented:
@cchaos Thanks for bringing up all these points. I hadn't fully considered the risk of inconsistencies in the use of callouts and I can understand the concern. To answer your question, I think the need for a "full width" callout in a flyout does not warrant the risk of inconsistent callout use. Having said that though I do think we should allow for a "full width" banner in flyouts, it just doesn't have to mean be add a borderless prop to callout.

The flyout with a regular callout as banner looks like this:

![image](https://user-images.githubusercontent.com/4016496/74054105-29967300-4992-11ea-823f-7215532ce4b8.png)

I don't think this looks too bad which makes me wonder if we'd need to remove the border at all. We could always leave it up to the consumer to remove the border in a callout when using it in a flyout? Is that bad practice?

cchaos commented:
> I don't think this looks too bad which makes me wonder if we'd need to remove the border at all. 

I'll agree that it doesn't look too "bad", but obviously the non-bordered looks a bit better. I'll leave it up to you to decide if you'd want to keep the border in the flyout or not. However, I'd just not add it as a direct option of EuiCallout, but a custom override with classes.

> We could always leave it up to the consumer to remove the border in a callout when using it in a flyout? Is that bad practice?

By leave it up to the consumer, do you mean letting them do the overrides or using a prop on EuiCallout? Consumers can override anything they want really, we just don't usually add docs around this specifically. However, adding this override into EUI would be as simple as:

- Wrapping the `banner` element in a div with a custom class like: 
```jsx
<div className="euiFlyout__banner">{banner}</div>`. 
```

- In the css, watch for `.euiCallOut` as being a child of `.euiFlyout__banner` and then remove the banner if so:
```scss
.euiFlyout__banner .euiCallOut {
  border: none;
}
```

andreadelrio commented:
@cchaos Agreed about not adding borderless prop to callouts anymore. Yeah I know consumers can override anything, I guess I was wondering if it'll be weird for them to pass a "regular" callout as banner and see it doesn't have a border. But I guess that's not a big deal and I could always make a note in the docs that callouts get their border removed in flyouts.

----

> If we do move forward with this, something to consider is that the overflow mask in the flyout body is causing the top portion of the callout to fade out as it approaches the header border.
> 
> <img alt="Screen Shot 2020-02-07 at 08 46 55 AM" width="584" src="https://user-images.githubusercontent.com/549577/74034503-68bec700-4986-11ea-838f-62f70469aba6.png">

About this, I also noticed it and wasn't sure how to handle it. Would we push the banner down a bit, 

![image](https://user-images.githubusercontent.com/4016496/74060369-a92a3f00-499e-11ea-84d6-a7d358c1c35c.png)

or remove `mask-image`?

![image](https://user-images.githubusercontent.com/4016496/74060318-931c7e80-499e-11ea-9f32-e05a59111415.png)

I like how it looks without mask-image better but don't know how strict we are when it comes to allowing for this type of modifications. Or is there another way to handle it?

cchaos commented:
I would consider creating an alternate `mask-image` that just masks the bottom portion and use that.

andreadelrio commented:
> I would consider creating an alternate `mask-image` that just masks the bottom portion and use that.

@cchaos My plan for this would be to modify [euiOverflowShadow](https://github.com/elastic/eui/blob/master/src/global_styling/mixins/_shadow.scss#L89)  to also take a parameter called `type` which would default to `both`, but could also take `top` and `bottom`. We'd use it with `bottom` for this case. Let me know if this sounds like a good idea so I can proceed.

andreadelrio commented:
@cchaos I made the changes we discussed. I also updated the docs to reflect the changes on `euiOverflowShadow`. I did that for vertical scrolling, not sure if I should do it for horizontal scrolling too as I don't want the portion of the docs to get repetitive

![image](https://user-images.githubusercontent.com/4016496/74285001-c3835600-4cd9-11ea-8aaf-62e9d43b4232.png)

Updated flyout with banner and bottom shadows:

![image](https://user-images.githubusercontent.com/4016496/74285142-12c98680-4cda-11ea-95f7-ed1ca5b54796.png)

cchaos commented on src/components/flyout/flyout_body.tsx:
```diff
@@ -1,21 +1,34 @@
+  HTMLAttributes<HTMLDivElement> &
+    CommonProps & {
+      /**
+       * Use to display a banner at the top of the body. It is suggested to use a borderless `EuiCallOut` for it.
```
May want to adjust this comment now that there's no such thing as a borderless callout.

cchaos commented on src/components/flyout/_flyout_body.scss:
```diff
@@ -5,6 +5,19 @@
+  }
+
+  .euiFlyoutBody__overflow-banner .euiCallOut {
+    border: none;
```
Add a comment in here why you're altering a different component

cchaos commented on src/components/flyout/flyout_body.tsx:
```diff
@@ -1,21 +1,34 @@
 }) => {
   const classes = classNames('euiFlyoutBody', className);
+  const overflowClasses = classNames('euiFlyoutBody__overflow', {
+    'euiFlyoutBody__overflow-banner': banner,
```
Remember that BEM naming suggests that modifiers should use a `--` and you might want to say it `hasBanner` instead otherwise it sounds like it's trying to target the banner itself.

This class also shouldn't be the one used to target the callout or else any callouts within the rest of `children` will also get the border removed. So instead, your element would be:

```tsx
// If banner exists (= true), then spit out the wrapper plus the banner element
{banner &&
  <div class="euiFlyoutBody__banner">
    {banner}
  </div>
}
```

cchaos commented on src/global_styling/mixins/_shadow.scss:
```diff
@@ -86,12 +86,25 @@
+      transparentize(red, .9) 100%;
+    }
+  } @else {
+    @warn "euiOverflowShadow() expects side to be 'both', 'start' or 'end' but got '#{$side}'";
```
Nice! 🎉

cchaos commented on src/global_styling/mixins/_shadow.scss:
```diff
@@ -86,12 +86,25 @@
+    } @else {
+      $gradient: transparentize(red, 0) calc(100% - #{$hideHeight}),
+      transparentize(red, .9) 100%;
+    }
```
If you want to further reduce some repetition here. You could create two top variables of:

```scss
$gradientStart:
  transparentize(red, .9) 0%,
  transparentize(red, 0) $hideHeight;
$gradientEnd: 
  transparentize(red, 0) calc(100% - #{$hideHeight}),
  transparentize(red, .9) 100%;
```

And then use those in the if statement:

```scss
  @if ($side == 'both') {
      $gradient: $gradientStart, $gradientEnd;
    } @else if ($side == 'start') {
      $gradient: $gradientStart;
    } @else {
      $gradient: $gradientEnd;
    }
```

andreadelrio commented:
@cchaos I added the example with the banner. I also removed references to `onSwitchChange` and     
`isSwitchChecked` from the other examples as we are not using them, as far I can tell.
